test-runner/test_device_send_telemetry.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python
+
+# Copyright (c) Microsoft. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for
+# full license information.
+
+import pytest
+import connections
+import test_utilities
+import environment
+from adapters import print_message as log_message
+import concurrent
+import asyncio
+
+
+@pytest.mark.testgroup_iothub_module_client
+@pytest.mark.callsSendEvent
+def test_device_send_event_to_iothub():
+
+    log_message("connecting device client")
+    device_client = connections.connect_leaf_device_client()
+    log_message("connecting eventhub client")
+    eventhub_client = connections.connect_eventhub_client()
+
+    sent_message = test_utilities.random_string_in_json()
+    log_message("sending event: " + str(sent_message))
+    device_client.send_event(sent_message)
+
+    log_message("wait for event to arrive at eventhub")
+    received_message = eventhub_client.wait_for_next_event(
+        environment.leaf_device_id,
+        test_utilities.default_eventhub_timeout,
+        expected=sent_message,
+    )
+    if not received_message:
+        log_message("Message not received")
+        assert False
+
+    log_message("disconnecting device client")
+    device_client.disconnect()
+    log_message("disconnecting eventhub client")
+    eventhub_client.disconnect()
+
+
+@pytest.mark.asyncio
+@pytest.mark.testgroup_iothub_module_client
+@pytest.mark.callsSendEvent
+async def test_device_send_event_to_iothub_async():
+
+    executor = concurrent.futures.ThreadPoolExecutor()
+    loop = asyncio.get_event_loop()
+
+    log_message("connecting device client")
+    device_client_async = await connections.connect_leaf_device_client_async()
```

Comment:
Why does this test use leaf device client, when the module equivalent does not use a leaf module client (it uses a regular module client)?

Also, I'm not entirely clear on what a leaf device is. My understanding is that a leaf device is a limited device, used in receiving tests (i.e. can it receive a method call sent from elsewhere). Why is it being used to SEND an event here?

Reply:
yes as far as I know the only device client which was available in the framework was a leaf device. I just used whatever was available. I think we can introduce another device client though I need to discuss with Bert on this as well. I feel that other device client may not be usable for other SDKs.